libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -598,22 +597,81 @@ string YulUtilFunctions::clearStorageRangeFunction(Type const& _type)
 {
 	string functionName = "clear_storage_range_" + _type.identifier();
 
-	solUnimplementedAssert(_type.isValueType(), "...");
-
 	return m_functionCollector->createFunction(functionName, [&]() {
 		return Whiskers(R"(
 			function <functionName>(start, end) {
 				for {} lt(start, end) { start := add(start, 1) }
 				{
-					sstore(start, 0)
+					<setToZero>(start, 0)
 				}
 			}
 		)")
 		("functionName", functionName)
+		("setToZero", storageSetToZeroFunction(_type))
 		.render();
 	});
 }
 
+string YulUtilFunctions::clearStorageArrayFunction(ArrayType const& _type)
+{
+	solAssert(_type.location() == DataLocation::Storage, "");
+
+	if (_type.baseType()->storageBytes() < 32)
+	{
+		solAssert(_type.baseType()->isValueType(), "Invalid storage size for non-value type.");
+		solAssert(_type.baseType()->storageSize() <= 1, "Invalid storage size for type.");
+	}
+
+	if (_type.baseType()->isValueType())
+		solAssert(_type.baseType()->storageSize() <= 1, "Invalid size for value type.");
+
+	string functionName = "clear_storage_array_" + _type.identifier();
+
+	return m_functionCollector->createFunction(functionName, [&]() {
+		if (_type.isDynamicallySized())
+			return Whiskers(R"(
+				function <functionName>(slot) {
+					<clearDynamicArray>(slot)
+				}
+			)")
+			("functionName", functionName)
+			("clearDynamicArray", clearDynamicStorageArrayFunction(_type))
+			.render();
+
+		return Whiskers(R"(
+			function <functionName>(slot) {
+				<clearRange>(slot, add(slot, <lenToSize>(<len>)))
```

Comment:
I think the semantics of the clear range function is too confusing. Why don't you have to adjust the "end" parameter if you change the base type to uint256 "on the fly"? And why can you always increment by one in the loop, even if the base type is larger?

Reply:
It should be fine to limit the clear range function to full slots, because otherwise it gets to complicated if the number of items to clear does not end with a full slot. But then, there should be an assertion that the type is at least 17 bytes wide. The increment, though, should be the storage stride of the array.